daffodil-core/src/test/scala/org/apache/daffodil/dsom/TestRefMap.scala

```diff
@@ -0,0 +1,651 @@
+        </xs:complexType>
+      </xs:element>)
+    val compiler = Compiler()
+    val sset = compiler.compileNode(testSchema).sset
+    val root = sset.root
+    val comps = root.allComponents
+    assertEquals(253, comps.length)
+    val refMap = root.refMap
+    val numEntries = refMap.size
+    assertEquals(6, numEntries)
+    val refPairsMap = root.refPairsMap.toSeq
+    val numRefPairs = refPairsMap.map { _._2.length }.sum
+    assertEquals(11, numRefPairs)
+    assertEquals(253, root.numComponents)
+    assertEquals(25, root.numUniqueComponents)
+  }
+
+  @Test def testRefMapExplosion6() {
+    val testSchema = SchemaUtils.dfdlTestSchema(
+      <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>,
+      <dfdl:format ref="tns:GeneralFormat"/>,
+      <xs:element name="e7">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e6"/>
+            <xs:element ref="ex:e6"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e6">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e5"/>
+            <xs:element ref="ex:e5"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e5">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e4"/>
+            <xs:element ref="ex:e4"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e4">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e3"/>
+            <xs:element ref="ex:e3"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e3">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e2"/>
+            <xs:element ref="ex:e2"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e2">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element ref="ex:e1"/>
+            <xs:element ref="ex:e1"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+      <xs:element name="e1">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element name="e0" type="xs:int" dfdl:length="1" dfdl:lengthKind="explicit"/>
+            <xs:element name="f0" type="xs:int" dfdl:length="1" dfdl:lengthKind="explicit"/>
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>)
+    val compiler = Compiler()
+    val sset = compiler.compileNode(testSchema).sset
+    val root = sset.root
+    val comps = root.allComponents
+    assertEquals(509, comps.length)
+    val refMap = root.refMap
+    val numEntries = refMap.size
+    assertEquals(7, numEntries)
+    val refPairsMap = root.refPairsMap.toSeq
+    val numRefPairs = refPairsMap.map { _._2.length }.sum
+    assertEquals(13, numRefPairs)
+    assertEquals(509, root.numComponents)
```

Comment:
If you look at the last 2 lines of most of these tests. They show the growth of components vs. unique components.